components/raftstore-v2/src/operation/ready/snapshot.rs

```diff
@@ -0,0 +1,275 @@
+    /// Validate the snapshot. Returns true if it's valid.
+    fn validate_snap(&self, snap: &Snapshot, request_index: u64) -> bool {
+        let idx = snap.get_metadata().get_index();
+        // TODO(nolouch): check tuncated index
+        if idx < request_index {
+            // stale snapshot, should generate again.
+            info!(
+                self.logger(),
+                "snapshot is stale, generate again";
+                "snap_index" => idx,
+                "request_index" => request_index,
+            );
+            STORE_SNAPSHOT_VALIDATION_FAILURE_COUNTER.stale.inc();
+            return false;
+        }
+
+        let mut snap_data = RaftSnapshotData::default();
+        if let Err(e) = snap_data.merge_from_bytes(snap.get_data()) {
+            error!(
+                self.logger(),
+                "failed to decode snapshot, it may be corrupted";
+                "err" => ?e,
+            );
+            STORE_SNAPSHOT_VALIDATION_FAILURE_COUNTER.decode.inc();
+            return false;
+        }
+        let snap_epoch = snap_data.get_region().get_region_epoch();
+        let latest_epoch = self.region().get_region_epoch();
+        if snap_epoch.get_conf_ver() < latest_epoch.get_conf_ver() {
+            info!(
+                self.logger(),
+                "snapshot epoch is stale";
+                "snap_epoch" => ?snap_epoch,
+                "latest_epoch" => ?latest_epoch,
+            );
+            STORE_SNAPSHOT_VALIDATION_FAILURE_COUNTER.epoch.inc();
+            return false;
+        }
+
+        true
+    }
+
+    /// Cancel generating snapshot.
+    pub fn cancel_generating_snap(&self, compact_to: Option<u64>) {
+        let mut snap_state = self.snap_state_mut();
+        let SnapState::Generating {
+           ref canceled,
+           ref index,
+        } = *snap_state else { return };
+
+        if let Some(idx) = compact_to {
+            let snap_index = index.load(Ordering::SeqCst);
+            if snap_index == 0 || idx <= snap_index + 1 {
+                return;
+            }
+        }
+        canceled.store(true, Ordering::SeqCst);
+        *snap_state = SnapState::Relax;
+        info!(
+            self.logger(),
+            "snapshot is canceled";
+            "compact_to" => compact_to,
+        );
+        STORE_SNAPSHOT_VALIDATION_FAILURE_COUNTER.cancel.inc();
+    }
+
+    /// Try to switch snap state to generated. only `Generating` can switch to
+    /// `Generated`.
+    ///  TODO: make the snap state more clearer, the snapshot must be consumed.
+    pub fn try_switch_snap_state_to_generated(&self, snap: Box<Snapshot>) -> bool {
+        let mut snap_state = self.snap_state_mut();
+        let SnapState::Generating {
+            ref canceled,
+            ref index,
+         } = *snap_state else { return false };
+
+        if snap.get_metadata().get_index() < index.load(Ordering::SeqCst) {
+            drop(snap_state);
+            self.cancel_generating_snap(None);
```

Comment:
Why cancel?

Reply:
the snap index is invalid here. but I think it impossible to happen. I see you commented to check the index before, what do you mean?

Reply:
It can happen in following sequence:

0. raft calls snapshot
1. snapshot is scheduled and generated
2. snapshot is canceled
3. raft calls snapshot again
4. snapshot is scheduled and generated
5. peer storage on snapshot generated with old snapshot generated at 1.

I don't think it needs to be canceled as it's still waiting for snapshot generated at 4.

Reply:
got, I removed cancel here.